pykokkos/core/translators/functor.py

```diff
@@ -147,6 +147,50 @@ def generate_constructor(
 
     return cppast.ConstructorDecl("", name, params, body)
 
+def generate_constructor_without_rand(
+    name: str,
+    fields: Dict[cppast.DeclRefExpr, cppast.PrimitiveType],
+    views: Dict[cppast.DeclRefExpr, cppast.ClassType]
+) -> cppast.ConstructorDecl:
+    """
+    Generate the functor constructor
+
+    :param name: the functor class name
+    :param fields: a dict mapping from field name to type
+    :param views: a dict mapping from view name to type
+    :param random_pool: a tuple of the pool name and type
+    :param has_rand_call: whether the function contains a call to pk.rand()
+    :returns: the cppast representation of the constructor
+    """
+
+    params: List[cppast.ParmVarDecl] = []
+    assignments: List[cppast.AssignOperator] = []
+
+    for n, t in fields.items():
+        params.append(cppast.ParmVarDecl(t, n))
+
+    for n, t in views.items():
+        # skip subviews
+        if t is None:
+            continue
+        view_type: str = get_view_type(t)
+        params.append(cppast.ParmVarDecl(view_type, n))
+
+    # Kokkos fails to compile a functor if there are no parameters in its constructor
+    if len(params) == 0:
+        decl = cppast.DeclRefExpr("pk_field")
+        type = cppast.PrimitiveType(cppast.BuiltinType.INT)
+        params.append(cppast.ParmVarDecl(type, decl))
+
+    assignments.extend(generate_assignments(fields))
+    # skip subviews
+    assignments.extend(generate_assignments({v: views[v] for v in views if views[v]}))
+
+    body = cppast.CompoundStmt(assignments)
+
+    return cppast.ConstructorDecl("", name, params, body)
+
+
```

Comment:
I wonder if it's possible to refactor this so that we reuse `generate_constructor()`? Kind of like what you did in bindings.py for `generate_copy_back()`. Just add an extra boolean (e.g. `add_rand_params`) that guards the addition of the random pool parameters to the constructor.

Reply:
I was trying ... but the rationale for not doing it was that it would be a major change to the functors interface. In the state before this pr, the constructor of the functor was getting the rand arguments no matter what ... if they were not used in the constructor they would go out of scope. But everything in pykokkos now expects the functor to be constructible this way.
In c++ this kind of is a problem, as the random arguments alone make the functor dependent on the types that are the random_args even if they are never used in the constructor implementation. This basically ties the type of the functor to other types (seen by having to include `<Kokkos_Random.hpp>`) ... and any unnecessary dependencies in code will make future maintenance harder. Just think about that a change in `<Kokkos_Random.hpp>` might change the compile flags necessary ... and thus we would need to adapt both compilation and linking even though we might never use anything from the header. 
Thus I think in the long run, the constructor should only get the random args if it really uses them.
Even though this would be something I could do in this pr, it would be a different topic and I would rather do it in another pr (at best the two constructors that now do almost the same thing would be changed so that one uses the other so we do not repeat the implementation)
Furthermore, we might be interested in thinking about the functor constructors in general, as they require every member to be default constructible without actually needing to require this.
Additionally, we do a bunch of copies in the wrappers and the constructor that could be avoided. 

I guess the tl:dr is:
I think we should think about refactoring the functor creation in a broader context which deserves it's own pr. We could discuss this in the pykokkos meeting or in an issue